### PR TITLE
multi_nics_verify: Boot 27 NICs and all the NICs can get IPs

### PR DESCRIPTION
**Pull request by xiywang**
id: 1581019

Signed-off-by: Xiyue Wang <xiywang@redhat.com>

### CODE REVIEW

**Comment by xiywang** [comment]
python ConfigTest.py --testcase=multi_nics_verify.get_ips --guestname=RHEL.7.6,Win10 --platform=x86_64 --clone=no

 (1/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.myprovider.multi_nics_verify.get_ips: PASS (92.92 s)
 (2/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.myprovider.multi_nics_verify.get_ips: PASS (419.47 s)

File: qemu/tests/multi_nics_verify.py [hunk 1]
```diff
@@ -35,39 +42,66 @@ def check_nics_num(expect_c, session):
...
+            guest_ip = utils_net.get_guest_ip_addr(session, nic.mac, os_type,
+                                                   ip_version="ipv4")
+            if not guest_ip:
+                err_log = "vm get interface %s's ip failed." % index
+                test.fail(err_log)
+            nic_interface.append(guest_ip)
+        nic_interface_len = len(nic_interface)
+        if nics_num != nic_interface_len:
+            test.fail("Not all the nics in the guest have ips!")
+        else:
```
Comment (sitoliu) [inline qemu/tests/multi_nics_verify.py]: No need to add `else`, `logging.info` is ok.

File: qemu/tests/multi_nics_verify.py [hunk 2]
```diff
@@ -35,39 +42,66 @@ def check_nics_num(expect_c, session):
...
-        eth_config = "DEVICE=%s\\nBOOTPROTO=dhcp\\nONBOOT=yes" % ifname
-        cmd = "echo -e '%s' > %s" % (eth_config, eth_config_path)
-        s, o = session.cmd_status_output(cmd)
-        if s != 0:
-            err_msg = "Failed to create ether config file: %s\nReason is: %s"
-            test.error(err_msg % (eth_config_path, o))
+    os_type = params.get("os_type", "linux")
+    if params.get("redirect_log_file", "yes") == "no":
+        # Check all the interfaces in guest get ips
+        guest_ifname = ""
```
Comment (sitoliu) [inline qemu/tests/multi_nics_verify.py]: This variable is not been used.

File: qemu/tests/multi_nics_verify.py [hunk 3]
```diff
@@ -35,39 +42,66 @@ def check_nics_num(expect_c, session):
...
+        guest_ip = ""
```
Comment (sitoliu) [inline qemu/tests/multi_nics_verify.py]: I think you can use this variable directly, no need to define it at begin.

File: qemu/tests/multi_nics_verify.py [hunk 4]
```diff
@@ -35,39 +42,66 @@ def check_nics_num(expect_c, session):
...
+            nic_interface.append(guest_ip)
+        nic_interface_len = len(nic_interface)
+        if nics_num != nic_interface_len:
+            test.fail("Not all the nics in the guest have ips!")
+        else:
+            logging.info("All the [ %s ] NICs get IPs." % nics_num)
+    else:
+        # Redirect ifconfig output from guest to log file
+        log_file = os.path.join(test.debugdir, "ifconfig")
+        ifconfig_output = session.cmd("ifconfig")
```
Comment (sitoliu) [inline qemu/tests/multi_nics_verify.py]: Do you mean `session.cmd_output()`? I am not sure if `session.cmd()` can get the command's output.

File: qemu/tests/multi_nics_verify.py (hunk 1 above)
Comment (xiywang) [reply qemu/tests/multi_nics_verify.py]: i think to keep the else or not is not so rigorous.
both method are fine.

File: qemu/tests/multi_nics_verify.py (hunk 4 above)
Comment (xiywang) [reply qemu/tests/multi_nics_verify.py]: session.cmd() can also works fine.

File: qemu/tests/multi_nics_verify.py (hunk 3 above)
Comment (xiywang) [reply qemu/tests/multi_nics_verify.py]: removed

**Comment by xiywang** [comment]
@sitoliu @pezhang Please help to review again. Thanks.

File: qemu/tests/multi_nics_verify.py [hunk 5]
```diff
@@ -35,39 +42,68 @@ def check_nics_num(expect_c, session):
...
+        ifcfg_path = "/etc/sysconfig/network-scripts/ifcfg-%s"
+        for ifname in ifname_list:
+            eth_config_path = ifcfg_path % ifname
+            eth_config = "DEVICE=%s\\nBOOTPROTO=dhcp\\nONBOOT=yes" % ifname
+            cmd = "echo -e '%s' > %s" % (eth_config, eth_config_path)
+            s, o = session.cmd_status_output(cmd)
+            if s != 0:
+                err_msg = "Failed to create ether config file: %s\nReason is: %s"
+                test.error(err_msg % (eth_config_path, o))
+        log_file_object.close()
```
Comment (sitoliu) [inline qemu/tests/multi_nics_verify.py]: You didn't use `log_file_object` in line67-line80, what about move this statement below line65?

File: qemu/tests/multi_nics_verify.py [hunk 6]
```diff
@@ -35,39 +42,68 @@ def check_nics_num(expect_c, session):
...
+
+        # Reboot and check the configurations.
+        new_session = vm.reboot(session)
+        s, msg = check_nics_num(nics_num, new_session)
+        if not s:
+            test.fail(msg)
+        session.close()
```
Comment (sitoliu) [inline qemu/tests/multi_nics_verify.py]: s/session/new_session

File: qemu/tests/multi_nics_verify.py [hunk 7]
```diff
@@ -35,39 +42,68 @@ def check_nics_num(expect_c, session):
...
+        logging.info("index %s nic" % index)
+        guest_ip = utils_net.get_guest_ip_addr(session_srl, nic.mac, os_type,
+                                               ip_version="ipv4")
+        if not guest_ip:
+            err_log = "vm get interface %s's ip failed." % index
+            test.fail(err_log)
+        nic_interface.append(guest_ip)
+    nic_interface_len = len(nic_interface)
+    session_srl.close()
+    if nics_num != nic_interface_len:
```
Comment (sitoliu) [inline qemu/tests/multi_nics_verify.py]: You have already raise `test.fail` in line102 when nic doesn't get an ip address.

**Comment by pezhang** [comment]
(1/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.multi_nics_verify: PASS (349.86 s)

 (2/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.multi_nics_verify: PASS (424.46 s)

File: qemu/tests/multi_nics_verify.py (hunk 7 above)
Comment (xiywang) [reply qemu/tests/multi_nics_verify.py]: @sitoliu @pezhang 
Updated. Please help review again. Thanks.

**Comment by sitoliu** [comment]
LGTM.
Acked-by: Sitong Liu <siliu@redhat.com>

**Comment by xiywang** [comment]
@sitoliu Updated.

**Comment by pezhang** [comment]
Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.multi_nics_verify: PASS (421.46 s)

 Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.multi_nics_verify: PASS (427.56 s)

Looks good to me.
Acked-by: Pei Zhang pezhang@redhat.com

File: qemu/tests/multi_nics_verify.py [hunk 8]
```diff
@@ -35,39 +42,66 @@ def check_nics_num(expect_c, session):
...
     vm = env.get_vm(params["main_vm"])
     vm.verify_alive()
     session = vm.wait_for_login(timeout=int(params.get("login_timeout", 360)))
-    # Redirect ifconfig output from guest to log file
-    log_file = os.path.join(test.debugdir, "ifconfig")
-    ifconfig_output = session.cmd("ifconfig")
-    log_file_object = open(log_file, "w")
-    log_file_object.write(ifconfig_output)
 
     # Get the ethernet cards number from params
```
Comment (vivianQizhu) [inline qemu/tests/multi_nics_verify.py]: Seems this comment should be moved before L#45?

File: qemu/tests/multi_nics_verify.py [hunk 9]
```diff
@@ -35,39 +42,66 @@ def check_nics_num(expect_c, session):
...
-    nics_num = len(params.objects("nics"))
     logging.info("[ %s ] NICs card specified in config file" % nics_num)
```
Comment (vivianQizhu) [inline qemu/tests/multi_nics_verify.py]: So does this line.

File: qemu/tests/multi_nics_verify.py (hunk 6 above)
Comment (vivianQizhu) [reply qemu/tests/multi_nics_verify.py]: Or move `session.close()` to L#79

File: qemu/tests/multi_nics_verify.py [hunk 10]
```diff
@@ -35,39 +42,66 @@ def check_nics_num(expect_c, session):
...
+    session_srl = vm.wait_for_serial_login(timeout=int(params.get("login_timeout", 360)))
+    for index, nic in enumerate(vm.virtnet):
+        logging.info("index %s nic" % index)
+        guest_ip = utils_net.get_guest_ip_addr(session_srl, nic.mac, os_type,
+                                               ip_version="ipv4")
+        if not guest_ip:
+            err_log = "vm get interface %s's ip failed." % index
+            test.fail(err_log)
+        nic_interface.append(guest_ip)
+    nic_interface_len = len(nic_interface)
```
Comment (vivianQizhu) [inline qemu/tests/multi_nics_verify.py]: Is this `nic_interface_len` useless? Or you want to compare it with `nics_num`?

File: qemu/tests/multi_nics_verify.py (hunk 6 above)
Comment (xiywang) [reply qemu/tests/multi_nics_verify.py]: Should not at line 79. This is a for loop. Moved to line 82.

**Comment by xiywang** [comment]
@vivianQizhu Updated. Please help review again. Thanks.

**Pull request merged**